lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -81,12 +81,46 @@ openModuleFiles(StringRef DirName, StringRef ModuleFilename,
   return std::error_code();
 }
 
+static void addDiagnosticInfoForArchitectureMismatch(ASTContext &ctx,
+                             SourceLoc sourceLocation,
+                             llvm::SmallString<64> moduleName,
+                             llvm::SmallString<16> archName,
+                             llvm::SmallString<128> directoryPath) {
+
+  std::error_code errorCode;
+  llvm::Twine twineDirPath(directoryPath);
+  llvm::sys::fs::directory_iterator directoryIterator(twineDirPath,
+                                                      errorCode,
+                                                      true);
+  llvm::sys::fs::directory_iterator endIterator;
+
+  if(errorCode) {
+      return;
+    }
+
+  std::string foundArchs;
+  for (; directoryIterator != endIterator; directoryIterator.increment(errorCode)) {
```

Comment:
You should probably stop the loop if there's ever an error, too, even if it means having to emit a more generic diagnostic.